src/graph/ford_fulkerson.rs

```diff
@@ -1,141 +1,311 @@
+#[derive(Debug, PartialEq)]
+pub enum FordFulkersonError {
+    /// Error indicating that the graph is empty or has no edges.
+    EmptyGraph,
+    /// Indicates that the graph is not a square matrix.
+    ImproperGraph,
+    /// Error indicating that the source vertex is out of bounds.
+    SourceOutOfBounds,
+    /// Error indicating that the sink vertex is out of bounds.
+    SinkOutOfBounds,
+}
 
-pub fn bfs(r_graph: &[Vec<i32>], s: usize, t: usize, parent: &mut [i32]) -> bool {
-    let mut visited = [false; V];
-    visited[s] = true;
-    parent[s] = -1;
+/// Performs a Breadth-First Search (BFS) on the residual graph to find an augmenting path
+/// from the source vertex `source` to the sink vertex `sink`.
+///
+/// # Arguments
+///
+/// * `graph` - A mutable reference to the residual graph represented as an adjacency matrix.
+/// * `source` - The source vertex.
+/// * `sink` - The sink vertex.
+/// * `parent` - A mutable reference to the parent array used to store the augmenting path.
+///
+/// # Returns
+///
+/// Returns `true` if an augmenting path is found from `source` to `sink`, `false` otherwise.
+fn bfs(graph: &[Vec<isize>], source: usize, sink: usize, parent: &mut [isize]) -> bool {
+    let mut visited = vec![false; graph.len()];
+    visited[source] = true;
+    parent[source] = -1;
 
     let mut queue = VecDeque::new();
-    queue.push_back(s);
-
-    while let Some(u) = queue.pop_front() {
-        for v in 0..V {
-            if !visited[v] && r_graph[u][v] > 0 {
-                visited[v] = true;
-                parent[v] = u as i32; // Convert u to i32
-                if v == t {
+    queue.push_back(source);
+
+    while let Some(current_vertex) = queue.pop_front() {
+        for (previous_vertex, &capacity) in graph[current_vertex].iter().enumerate() {
+            if !visited[previous_vertex] && capacity > 0 {
+                visited[previous_vertex] = true;
+                parent[previous_vertex] = current_vertex as isize;
+                if previous_vertex == sink {
                     return true;
                 }
-                queue.push_back(v);
+                queue.push_back(previous_vertex);
             }
         }
     }
 
     false
 }
 
-pub fn ford_fulkerson(graph: &[Vec<i32>], s: usize, t: usize) -> i32 {
-    let mut r_graph = graph.to_owned();
-    let mut parent = vec![-1; V];
+/// Applies the Ford-Fulkerson algorithm to find the maximum flow in a flow network
+/// represented by a weighted directed graph.
+///
+/// # Arguments
+///
+/// * `graph` - A mutable reference to the flow network represented as an adjacency matrix.
+/// * `source` - The source vertex.
+/// * `sink` - The sink vertex.
+///
+/// # Returns
+///
+/// Returns the maximum flow and the residual graph
+pub fn ford_fulkerson(
+    graph: &[Vec<isize>],
+    source: usize,
+    sink: usize,
+) -> Result<isize, FordFulkersonError> {
+    if graph.is_empty() || graph[0].is_empty() {
```

Comment:
I am not sure about this check, i.e. about the `graph[0].is_empty()`. If `graph[0]` exists, then `graph.len() > 0`, so if we only deal with _proper_ graphs (i.e. having square _flow_ matrix), then `graph[0].len()` must be `graph.len()`.

If you remove this part of the condition, please update the comment explaining the `FordFulkersonError::EmptyGraph`.

Reply:
If the graph is a square matrix, which it should be for a proper flow network, then `graph[0].len()` will always be equal to `graph.len()` if `graph` is non-empty. As a result, the check for `graph[0].is_empty()` is unnecessary. We can simplify the condition and update the error handling accordingly